### PR TITLE
Departmental officers are now put together, rather than separate, if possible

### PR DESCRIPTION
**Pull request by Mothblocks**
## About The Pull Request
Security officers will now be paired up together in the same department, across departments. This means that, instead of 4 officers being split across 4 departments, there'll now be 2 groups of 2.

Late-join officers will be put into any department with only 1 officer. If none exist, the least populous department will be chosen, with their preference having priority.

Updates the maps to have more spawns for departmental officers. Delta previously had none, and now has some. Fixed a bug where MetaStation's security departmental officer was a supply one instead.

Removes the "random" departmental preference. All security officers are now given a department. The "none" preference still exists, but just to show that you don't care which department you're put into.

Updates the config to comment out SEC_START_BRIG. This is what the configuration was already on live servers.

This is something that should likely be test merged, but it mucks with savefiles. Luckily, the only damage it does is changing random departments to none, so if a TM is reverted, only that will have to be changed.

TODO:
- [x] Headset and PDA message to security officers about latejoiners being added to their department.
- [x] Only player sec officers should be assigned to departments.

## Why It's Good For The Game
As a security officer, your survival and fun are greatly increased when pairing up with a fellow security officer. I told @ninjanomnom that I believe that departmental security officers present a weird tacit acceptance of splitting up, however. You're essentially encouraged to go your own way. However, I'm not opposed to them either, as I think it can promote teamwork, as you get more access the more separate officers team up.

My solution is to try and encourage, through design, teaming up, while preserving departmental officers. Thus, they now pair up, which is, I believe, a good way to suggest to players to stick by each other. ninjanomnom was interested to see the results.

## Changelog
:cl:
add: Security officers are now paired up into departments together, rather than separated.
add: DeltaStation now has departmental security officer spawners.
fix: Fixed MetaStation's science officer being marked as supply.
/:cl:

### CODE REVIEW

**Comment by TheVekter** [comment]
That's a pretty salient point - sec does tend to be more enjoyable with other people. Does this make it so no sec officers naturally spawn in the brig? If so, we need to make sure each of the offices has enough gear for two people at roundstart.

**Comment by Mothblocks** [comment]
> Does this make it so no sec officers naturally spawn in the brig? If so, we need to make sure each of the offices has enough gear for two people at roundstart.

Correct, no officers naturally spawn in the brig.

You bring up a good point, but I'm hesitant about supplying the rooms themselves, as they're prime targets for antags/tiders to try and score loot like sec HUDs. That's not a bad thing, but it makes me uneasy to supply them more.

**Comment by TheVekter** [comment]
I recall sec not getting HUDs on round start as a purposeful balance thing, but I can't for the life of me remember _why_.

Either way, you could just put optional stuff like flashbangs, mace, etc. in there and leave out anything they spawn with.

**Comment by Mothblocks** [comment]
I'm going to opt out of that now for the sake of scope, that's a separate balance change. I do understand the concern, though.

**Comment by Dorsisdwarf** [comment]
> 
> 
> I'm going to opt out of that now for the sake of scope, that's a separate balance change. I do understand the concern, though.
> 
> > I recall sec not getting HUDs on round start as a purposeful balance thing, but I can't for the life of me remember why.
> 
> If you can find the source on this, that'd be appreciated! 😄

He's thinking of detectives IIRC

**Comment by Mickyan** [comment]
> You bring up a good point, but I'm hesitant about supplying the rooms themselves, as they're prime targets for antags/tiders to try and score loot like sec HUDs. That's not a bad thing, but it makes me uneasy to supply them more than they are now.

May be out of scope but I think it would be useful to have the ability to set different job loadouts between roundstart and latejoin, we have a bunch of equipment that's mapped in with the expectation that players will gear up at roundstart but we don't necessarily want every latejoin officer to be fully equipped right out of the arrivals shuttle

Giving a full kit only to roundstart officers would solve the problem of requiring them to make a beeline to the brig to gear up right away without adding a bunch of loot to every security checkpoint

**Comment by Mothblocks** [comment]
That's definitely possible, and sounds like a nice solution, though I will also consider that out of scope for now.

**Comment by TheVekter** [comment]
> 
> 
> > I'm going to opt out of that now for the sake of scope, that's a separate balance change. I do understand the concern, though.
> > > I recall sec not getting HUDs on round start as a purposeful balance thing, but I can't for the life of me remember why.
> > 
> > 
> > If you can find the source on this, that'd be appreciated! 😄
> 
> He's thinking of detectives IIRC

I was, thanks.

**Comment by ATH1909** [comment]
nooooo how will I always roll medsec or engisec now??????

**Comment by Mothblocks** [comment]
ath's greatest nightmare, he'll have to be with other people :(

**Comment by Mothblocks** [comment]
why did i say that? did ath make a comment and then delete it

**Comment by Mothblocks** [comment]
I've been told that security officers need to go to the brig in order to get their security belts already, so I don't think much will change on this front.

File: code/_globalvars/lists/flavor_misc.dm [hunk 1]
```diff
@@ -146,7 +146,7 @@ GLOBAL_LIST_INIT(ai_core_display_screens, sortList(list(
 			return "ai-portrait" //just take this until they decide
 		return "ai-[lowertext(input)]"
 
-GLOBAL_LIST_INIT(security_depts_prefs, sortList(list(SEC_DEPT_RANDOM, SEC_DEPT_NONE, SEC_DEPT_ENGINEERING, SEC_DEPT_MEDICAL, SEC_DEPT_SCIENCE, SEC_DEPT_SUPPLY)))
+GLOBAL_LIST_INIT(security_depts_prefs, sortList(list(SEC_DEPT_NONE, SEC_DEPT_ENGINEERING, SEC_DEPT_MEDICAL, SEC_DEPT_SCIENCE, SEC_DEPT_SUPPLY)))
```
Comment (coiax) [inline code/_globalvars/lists/flavor_misc.dm:149]: Break this out into multiple lines.

Comment (Mothblocks) [reply code/_globalvars/lists/flavor_misc.dm:149]: hes gone mad were going back to 80 cols

File: code/modules/client/preferences_savefile.dm [hunk 2]
```diff
@@ -419,6 +419,8 @@ SAVEFILE UPDATING/VERSIONING - 'Simplified', or rather, more coder-friendly ~Car
 
 	READ_FILE(S["preferred_ai_core_display"], preferred_ai_core_display)
 	READ_FILE(S["prefered_security_department"], prefered_security_department)
+	if (!(prefered_security_department in GLOB.security_depts_prefs))
+		prefered_security_department = SEC_DEPT_NONE
```
[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/modules/client/preferences_savefile.dm:423]: In theory it should be a migration, but `prefered_security_department` isn't loaded in time.

I can mark it that this is from save file version 40, which means that once 40 is less than the define we use to completely wipe your data, this code can be removed.

**Comment by Cyberboss** [comment]
Looked at the code but it's late, I'm tired, and there's a lot of it. Sooo...

**Comment by Mothblocks** [comment]
I'll throw it up on Sybil later then.

File: code/modules/client/preferences_savefile.dm [hunk 3]
```diff
@@ -431,6 +431,15 @@ SAVEFILE UPDATING/VERSIONING - 'Simplified', or rather, more coder-friendly ~Car
...
 	READ_FILE(S["prefered_security_department"], prefered_security_department)
 
+	// This is the version when the random security department was removed.
+	// When the minimum is higher than that version, it's impossible for someone to have the "Random" department.
+	#if SAVEFILE_VERSION_MIN > 40
+	#warn The prefered_security_department check in preferences_savefile.dm is no longer necessary.
+	#endif
+
+	if (!(prefered_security_department in GLOB.security_depts_prefs))
+		prefered_security_department = SEC_DEPT_NONE
```
Comment (LemonInTheDark) [inline code/modules/client/preferences_savefile.dm:441]: Is this needed? can't you use `update_preferences` to convert everything over after you raise the version?

File: code/modules/jobs/job_types/security_officer.dm [hunk 4]
```diff
@@ -149,11 +205,228 @@ GLOBAL_LIST_INIT(available_depts, list(SEC_DEPT_ENGINEERING, SEC_DEPT_MEDICAL, S
...
+	 * [engineer, medical, medical, medical, medical, cargo]
+	 *
+	 * The distribution of this is [2, 2, 2], meaning there will be 3 departments chosen and they will have 2 each.
+	 * We order from most popular to least popular and get:
+	 * - medical: 4
+	 * - engineer: 1
+	 * - cargo: 1
+	 *
+	 * We need 2 to fill the first group. There are enough medical staff to do it. Thus, we take the first 2 medical staff
+	 * and update the output, making it now: [?, medical, medical, ?, ?, cargo].
```
Comment (LemonInTheDark) [inline code/modules/jobs/job_types/security_officer.dm:270]: Why isn't engineer in this list? am I misunderstanding something?

File: code/modules/jobs/job_types/security_officer.dm [hunk 5]
```diff
@@ -149,11 +205,228 @@ GLOBAL_LIST_INIT(available_depts, list(SEC_DEPT_ENGINEERING, SEC_DEPT_MEDICAL, S
...
+
+	return distribution
+
+/proc/get_new_officer_distribution_from_late_join(
+	preference,
+	list/departments,
+	list/distribution,
+)
+	/**
+	 * For late joiners, we're forced to put them in an alone department at some point.
```
Comment (LemonInTheDark) [inline code/modules/jobs/job_types/security_officer.dm:390]: Why not make this the autodoc? Same for the above comment I suppose

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
God this is some good documentation

**Comment by Mothblocks** [comment]
it didnt crash :tada:

File: code/modules/client/preferences_savefile.dm (hunk 3 above)
Comment (Mothblocks) [reply code/modules/client/preferences_savefile.dm:441]: No, `prefered_security_department` isn't loaded by then, it's loaded after. Changing the order of savefile initialization is something I'm a bit queasy about for something this small.

File: code/modules/jobs/job_types/security_officer.dm (hunk 4 above)
Comment (Mothblocks) [reply code/modules/jobs/job_types/security_officer.dm:270]: Typo, looks like.

```suggestion
	 * and update the output, making it now: [engineer, medical, medical, ?, ?, cargo].
```

File: code/modules/jobs/job_types/security_officer.dm (hunk 5 above)
Comment (Mothblocks) [reply code/modules/jobs/job_types/security_officer.dm:390]: These are implementation details, auto-docs are for consumers.

File: code/modules/client/preferences_savefile.dm (hunk 3 above)
Comment (LemonInTheDark) [reply code/modules/client/preferences_savefile.dm:441]: hhhhhhhhhhhhhhhhhhhhhhhhhhhhhh

**Comment by Cyberboss** [comment]
Bababooey

**Comment by MrMelbert** [comment]
nooo ihad testmerge feedback waiting for the round to end rEEEE

**Comment by MrMelbert** [comment]
fuck it
image
image

I was getting PDAs from every seccie that joined, not just seccies to my department. (I was assigned to medical this shift)

**Comment by Mothblocks** [comment]
Easy fix, I forgot to actually check for the department. I'll get to it when I can, but I'm in the middle of a large PR right now, so can't promise an ETA.

**Pull request merged**